### PR TITLE
Script side implementation for Brace Completion.

### LINKED ISSUE #1484: Visual Studio: Support "Automatic brace completion"
In Visual Studio if you go into the options for TypeScript(Tools > Options > Text Editor > TypeScript > General) the "Automatic brace completion" option is grayed out.

Automatic brace completion options grayed out  

I was wondering why this is and if there is any way to turn this on. 
  
  
**Visual Studio Version**: *Visual Studio Express 2013 for Web v12.0.31101.00 Update 4*  
**OS**: **Windows 8.1*

### PR DESCRIPTION
**Pull request by paulvanbrenk**
This needs updated Visual Studio components to work.

Fixes #1484

### CODE REVIEW

File: src/harness/fourslash.ts [hunk 1]
```diff
@@ -1844,6 +1844,31 @@ namespace FourSlash {
             });
         }
 
+        public verifyBraceCompletionAtPostion(negative: boolean, openingBrace: string) {
+
+            let charCode = 0x28; // '('
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Use `ts.CharacterCodes`.

File: src/services/services.ts [hunk 2]
```diff
@@ -7310,6 +7312,40 @@ namespace ts {
             return { newText: result, caretOffset: preamble.length };
         }
 
+        function isValidBraceCompletionAtPostion(fileName: string, position: number, openingBrace: number): boolean {
+
+            // '<' currently not supported, figuring out if we're in a Generic Type vs. a comparison is too 
+            // expensive to do during typing scenarios
+            // i.e. wether we're dealing with:
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: whether

File: src/services/services.ts [hunk 3]
```diff
@@ -7310,6 +7312,40 @@ namespace ts {
             return { newText: result, caretOffset: preamble.length };
         }
 
+        function isValidBraceCompletionAtPostion(fileName: string, position: number, openingBrace: number): boolean {
+
+            // '<' currently not supported, figuring out if we're in a Generic Type vs. a comparison is too
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: is

File: src/services/services.ts [hunk 4]
```diff
@@ -7310,6 +7312,40 @@ namespace ts {
...
+            if (isInJsxAttribute(sourceFile, position)) {
+                return false;
+            }
+
+            if (isInJsxElement(sourceFile, position)) {
+                return openingBrace === CharacterCodes.openBrace;
+            }
+
+            if (isInTemplateString(sourceFile, position)) {
+                return false;
```
Comment (DanielRosenwasser) [inline src/services/services.ts]: I don't understand, why wouldn't you want to complete a curly brace following a `${`?

File: src/services/shims.ts [hunk 5]
```diff
@@ -220,6 +220,13 @@ namespace ts {
          */
         getDocCommentTemplateAtPosition(fileName: string, position: number): string;
 
+        /**
+         * Returns JSON-encoded boolean to indicate whether we should support brace location
+         * at the current position.
+         * E.g. we don't want Brace Completion inside string-literals, comments, etc.
```
Comment (DanielRosenwasser) [inline src/services/shims.ts]: brace completion, no caps

File: src/services/utilities.ts [hunk 6]
```diff
@@ -436,6 +436,35 @@ namespace ts {
...
+            (token.kind === SyntaxKind.LessThanToken && token.parent.kind === SyntaxKind.JsxText))) {
+            return true;
+        }
+
+        return false;
+    }
+
+    export function isInTemplateString(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+        return token && (token.kind === SyntaxKind.NoSubstitutionTemplateLiteral || token.kind === SyntaxKind.TemplateHead || token.kind === SyntaxKind.TemplateMiddle || token.kind === SyntaxKind.TemplateTail) && position > token.getStart();
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: :-1: you have `isInsideTemplateLiteral` and `isTemplateLiteralKind`, use those.

File: src/services/utilities.ts [hunk 7]
```diff
@@ -436,6 +436,35 @@ namespace ts {
...
+        }
+
+        return false;
+    }
+
+    export function isInJsxElement(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+
+        if (token && (token.kind === SyntaxKind.JsxText ||
+            (token.kind === SyntaxKind.LessThanToken && token.parent.kind === SyntaxKind.JsxText))) {
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: What is the distinction of these situations? Leave a comment explaining each.

File: src/services/utilities.ts [hunk 8]
```diff
@@ -436,6 +436,35 @@ namespace ts {
...
+        }
+
+        if (token && token.kind === SyntaxKind.CloseBraceToken && token.parent.kind === SyntaxKind.JsxExpression) {
+            return true;
+        }
+
+        return false;
+    }
+
+    export function isInJsxElement(sourceFile: SourceFile, position: number) {
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: What do you mean *in*? Inside the opening tag? The closing tag? The children? Does this include `JsxSelfClosingElement`s?

File: src/services/utilities.ts [hunk 9]
```diff
@@ -436,6 +436,35 @@ namespace ts {
         return isInCommentHelper(sourceFile, position, /*predicate*/ undefined);
     }
 
+    export function isInJsxAttribute(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+        if (token && (token.kind === SyntaxKind.JsxAttribute || token.kind === SyntaxKind.JsxSpreadAttribute)) {
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: `JsxAttribute` and `JsxSpreadAttribute` are not a tokens, they are nodes, so do you have tests where you run into this?

File: src/services/utilities.ts [hunk 10]
```diff
@@ -436,6 +436,35 @@ namespace ts {
         return isInCommentHelper(sourceFile, position, /*predicate*/ undefined);
     }
 
+    export function isInJsxAttribute(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+        if (token && (token.kind === SyntaxKind.JsxAttribute || token.kind === SyntaxKind.JsxSpreadAttribute)) {
+            return true;
+        }
+
+        if (token && token.kind === SyntaxKind.CloseBraceToken && token.parent.kind === SyntaxKind.JsxExpression) {
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: Wouldn't the parent be a `JsxAttribute`?

File: src/services/services.ts (hunk 4 above)
Comment (paulvanbrenk) [inline src/services/services.ts]: We don't support brace completion in an existing brace completion session. This is a VS limitation, and the required additional effort doesn't seem worth it (at the moment).

Comment (DanielRosenwasser) [inline src/services/services.ts]: (as discussed offline)

File: src/services/utilities.ts (hunk 9 above)
Comment (paulvanbrenk) [inline src/services/utilities.ts]: Yeah don't actually need those

File: src/services/utilities.ts (hunk 10 above)
Comment (paulvanbrenk) [inline src/services/utilities.ts]: Nope, turns out it's not

File: src/harness/fourslash.ts [hunk 11]
```diff
@@ -1844,6 +1844,31 @@ namespace FourSlash {
...
+
+            let charCode = 0x28; // '('
+            switch (openingBrace) {
+                case "{": charCode = ts.CharacterCodes.openBrace; break;
+                case "[": charCode = ts.CharacterCodes.openBracket; break;
+                case "'": charCode = ts.CharacterCodes.singleQuote; break;
+                case '"': charCode = ts.CharacterCodes.doubleQuote; break;
+                case "`": charCode = ts.CharacterCodes.backtick; break;
+                case "<": charCode = ts.CharacterCodes.lessThan; break;
+            }
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Consider using a `Map<CharacterCodes>` and writing

```ts
const openBraceMap: Map<CharacterCodes> = {
    "{": CharacterCodes.openBrace,
     // etc.
}
```

Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: You should error when the test-writer accidentally gave the wrong input.

File: src/harness/fourslash.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: Just make perform the assignment in the switch to make it more uniform

File: src/services/utilities.ts [hunk 12]
```diff
@@ -436,6 +436,37 @@ namespace ts {
...
+    /**
+     * returns true if the position is in between the open and close elements of an JSX expression.
+     */
+    export function isInsideJsxElement(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+
+        // make a distinction if we're dealing with an empty expression or a non-empty expression i.e
+        // <div>|</div vs <div>Hello | World</div> , where | is current cursor
+        if (token && (token.kind === SyntaxKind.JsxText ||
+            (token.kind === SyntaxKind.LessThanToken && token.parent.kind === SyntaxKind.JsxText))) {
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: Can you elaborate on why you need to do this in a comment (as discussed offline)?

File: src/services/utilities.ts [hunk 13]
```diff
@@ -436,6 +436,37 @@ namespace ts {
...
+            return true;
+        }
+
+        return false;
+    }
+
+    export function isInTemplateString(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+        return isTemplateLiteralKind(token.kind) && position > token.getStart();
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: pass `sourceFile` to `getStart`

File: src/services/services.ts [hunk 14]
```diff
@@ -7310,6 +7312,40 @@ namespace ts {
...
         }
 
+        function isValidBraceCompletionAtPostion(fileName: string, position: number, openingBrace: number): boolean {
+
+            // '<' is currently not supported, figuring out if we're in a Generic Type vs. a comparison is too 
+            // expensive to do during typing scenarios
+            // i.e. whether we're dealing with:
+            //      var x = new foo<| ( with class foo<T>{} )
+            // or 
+            //      var y = 3 <|
```
Comment (sandersn) [inline src/services/services.ts]: Why not always complete, even in the `var y = 3 < {` case? If it's just a typo, then the user will just backspace immediately, which should delete both the `{` and the `}`.

Comment (DanielRosenwasser) [inline src/services/services.ts]: That's a very unpleasant experience for if you're just trying to write `<`. You shouldn't make the user do more work to undo your mistakes in the language service.

Comment (sandersn) [inline src/services/services.ts]: Discussed with Paul offline. This is for completion of '<' with '>'. I thought it was for completion of '{' with '}'.

File: src/harness/fourslash.ts [hunk 15]
```diff
@@ -1844,6 +1844,37 @@ namespace FourSlash {
             });
         }
 
+        public verifyBraceCompletionAtPostion(negative: boolean, openingBrace: string) {
+
+            const openBraceMap: ts.Map<ts.CharacterCodes> = {
```
Comment (DanielRosenwasser) [inline src/harness/fourslash.ts]: For every test you're going to be creating this, so make this a static property

File: src/services/utilities.ts [hunk 16]
```diff
@@ -437,13 +437,53 @@ namespace ts {
...
+     */
+    export function isInsideJsxElementOrAttribute(sourceFile: SourceFile, position: number) {
+        let token = getTokenAtPosition(sourceFile, position);
+
+        if (!token) {
+            return false;
+        }
+
+        // <div>Hello |</div>
+        if (token.kind === SyntaxKind.LessThanToken && token.parent.kind === SyntaxKind.JsxText) {
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: Still doesn't make sense to me that the token is considered a child of the JsxText. That should be the child of the closing JSX tag. So I can't imagine how you're hitting this.

File: src/services/utilities.ts [hunk 17]
```diff
@@ -437,13 +437,53 @@ namespace ts {
...
+            return true;
+        }
+
+        // <div> { 
+        // |
+        // } < /div>
+        if (token.kind === SyntaxKind.LessThanToken && token.parent.kind === SyntaxKind.JsxExpression) {
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: The comment here doesn't seem to reflect the logic? When does a less-than token have anything to do with the JSX Expression?

File: src/services/utilities.ts [hunk 18]
```diff
@@ -437,13 +437,53 @@ namespace ts {
...
+            return true;
+        }
+
+        // <div> { |  or <div a={|
+        if (token && token.kind === SyntaxKind.CloseBraceToken && token.parent.kind === SyntaxKind.JsxExpression) {
```
Comment (DanielRosenwasser) [inline src/services/utilities.ts]: Where's the close brace in this comment?

Comment (paulvanbrenk) [inline src/services/utilities.ts]: there is none... that's the point.

File: src/services/utilities.ts (hunk 16 above)
Comment (RyanCavanaugh) [inline src/services/utilities.ts]: Agree, this should be unreachable. Log a bug with a repro if this line gets hit.

Comment (paulvanbrenk) [inline src/services/utilities.ts]: This is the case where that happens at marker #1

```ts
declare var react: any;

 var x = <div a="/*2*/" b={/*3*/}>
 /*1*/
 </div>;
```

**Comment by DanielRosenwasser** [comment]
There is some very bizarre handling here, but I am giving this a :+1: because users can always turn this setting off. 

Can you open an infrastructure bug on what we've found here?

**Pull request merged**